src/node/historical_queries_adapter.cpp

```diff
@@ -189,13 +196,112 @@ namespace ccf::historical
+
+    // We start at the previous write to the latest (current) service info.
+    auto service = ctx.tx.template ro<Service>(Tables::SERVICE);
+    LOG_DEBUG_FMT("current identity: {}", service->get()->cert.str());
+
+    // Iterate until we find the most recent write to the service info that
+    // precedes the target seqno.
+    std::optional<ServiceInfo> hservice_info = service->get();
+    SeqNo i = -1;
+    do
+    {
+      // If no previous service info version available, e.g. after recovery of
+      // a 1.x service, we scan linearly.
+      i = hservice_info->previous_service_identity_version.value_or(i - 1);
+      LOG_DEBUG_FMT("previous write: {}", i);
+      auto hstate = state_cache.get_state_at(i, i);
+      if (!hstate)
+      {
+        LOG_DEBUG_FMT("Historical state at {} not available yet", i);
+        return std::nullopt;
+      }
+      auto htx = hstate->store->create_read_only_tx();
+      auto hservice = htx.ro<Service>(Tables::SERVICE);
+      hservice_info = hservice->get();
+    } while (i > target_seqno || (i > 1 && !hservice_info));
+
+    if (!hservice_info)
+    {
+      throw std::runtime_error("Failed to locate previous service identity");
+    }
+
+    return hservice_info;
+  }
+
+  bool get_service_endorsements(
+    auto& ctx,
+    ccf::historical::StatePtr& state,
+    AbstractStateCache& state_cache,
+    NetworkIdentitySubsystemInterface& niss)
+  {
+    const auto& network_identity = niss.get();
+
+    if (state && state->receipt && state->receipt->node_cert)
+    {
+      auto& receipt = *state->receipt;
+      LOG_DEBUG_FMT("receipt cert: {}", receipt.node_cert->str());
+
+      auto v = crypto::make_unique_verifier(*receipt.node_cert);
+      if (!v->verify_certificate(
+            {&network_identity->cert}, {}, /* ignore_time */ true))
+      {
+        // The current service identity does not endorse the node certificate in
+        // the receipt, so we search for the the most recent write to the
+        // service info table before the historical transaction ID to get the
+        // historical service identity.
+
+        auto opt_psi = find_previous_service_identity(ctx, state, state_cache);
+        if (!opt_psi)
+        {
+          return false;
+        }
+
+        auto hpubkey = crypto::public_key_pem_from_cert(
+          crypto::cert_pem_to_der(opt_psi->cert));
+
+        auto eit = service_endorsement_cache.find(hpubkey);
+        if (eit != service_endorsement_cache.end())
+        {
+          receipt.service_endorsements = eit->second;
+        }
+        else
+        {
+          auto endorsement = create_endorsed_cert(
+            hpubkey,
+            ReplicatedNetworkIdentity::subject_name,
+            {},
+            crypto::OpenSSL::to_x509_time_string(
+              crypto::OpenSSL::to_time_t(NULL)),
+            1 /* days valid */,
+            network_identity->priv_key,
+            network_identity->cert,
+            true);
```

Comment:
Does this mean the historical service certificate that's included in the receipt contains the same public key as the original service certificate, but with a different validity period (and maybe other attributes)?

Reply:
> I'm very flexible about the validity period though - it's not particularly useful in any case, because all of this is about historical state and therefore "expired by definition".

Is it better to use the same validity period (and other fields such as CN) for the new certificate then?